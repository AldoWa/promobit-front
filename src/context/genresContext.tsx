import { createContext, useCallback, useEffect, useState } from 'react';

import api from '../services/api';

interface GenresProviderProps {
  children: React.ReactNode;
}

interface Genre {
  id: number;
  name: string;
}

export interface GenresResponse {
  genres: Genre[];
}

export interface GenresContextData {
  genres: Genre[];
  handleGenre: (id: number, type: 'add' | 'remove') => void;
  genresIDsSelected: number[];
  resetGenresSelected: () => void;
}

export const GenresContext = createContext<GenresContextData>(
  {} as GenresContextData
);
export function GenresProvider({ children }: GenresProviderProps) {
  const [genres, setGenres] = useState<Genre[]>([]);
  const [genresIDsSelected, setGenresSelected] = useState<number[]>([]);

  useEffect(() => {
    const getGenres = async () => {
      const { data } = await api.get<GenresResponse>('genre/movie/list?language=pt')

      setGenres(data.genres)
    }

    getGenres()
  }, [])

  const addGenreSelected = useCallback((id: number) => {
    setGenresSelected([...genresIDsSelected, id])
  }, [genresIDsSelected])

  const removeGenreSelected = useCallback((id: number) => {
    const newGenres = genresIDsSelected.filter(genre => genre !== id)
    setGenresSelected(newGenres)
  }, [genresIDsSelected])

  const resetGenresSelected = useCallback(() => setGenresSelected([]), [])

  const handleGenre = useCallback((id: number, type: 'add' | 'remove') => {
    if (type === 'add') {
      addGenreSelected(id)
      return;
    }
    removeGenreSelected(id)
  }, [addGenreSelected, removeGenreSelected])


  return (
    <GenresContext.Provider value={{ genres, genresIDsSelected, handleGenre, resetGenresSelected }}>
      {children}
    </GenresContext.Provider>
  );
}
