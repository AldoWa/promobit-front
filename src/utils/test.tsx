import { RenderOptions, render } from '@testing-library/react'

import { ChakraProvider } from '@chakra-ui/react'

interface AllTheProvidersProps {
  children: React.ReactNode
}

const AllTheProviders = ({children}: AllTheProvidersProps) => {
  return (
    <ChakraProvider>
      {children}
    </ChakraProvider>
  )
}

const customRender = (ui:React.ReactElement, options?: RenderOptions) =>
  render(ui, {wrapper: AllTheProviders, ...options})

// eslint-disable-next-line react-refresh/only-export-components
export * from '@testing-library/react'

// override render method
export {customRender as render}