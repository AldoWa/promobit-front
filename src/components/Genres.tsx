import React, { useContext, useEffect } from 'react';

import { Box, Image, Text } from '@chakra-ui/react';

import Close from '../assets/close.svg';
import { GenresContext } from '../context/genresContext';


interface GenresProps {
  text: string;
  id: number;
  handleGenre: (id: number, type: 'add' | 'remove') => void;
}

const Genres: React.FC<GenresProps> = ({ text, id, handleGenre }) => {
  const [isSelected, setIsSelected] = React.useState(false);

  const { genresIDsSelected } = useContext(GenresContext)

  useEffect(() => {
    const isGenreSelected = genresIDsSelected.find(genre => genre === id)
    setIsSelected(!!isGenreSelected)
  }, [genresIDsSelected, id])

  const handleClick = () => {
    setIsSelected(!isSelected)
    if (!isSelected) {
      handleGenre(id, 'add')
      return;
    }
    handleGenre(id, 'remove')
  }

  if (isSelected) {
    return (
      <Box color='white' bg='orange.500'
        py={2}
        px={4}
        height={10}
        display='flex'
        justifyContent='center'
        alignItems='center'
        borderRadius={4}
        as="button"
        onClick={handleClick}
      >
        <Text fontSize='1rem' lineHeight='24px' textAlign='center'>
          {text}
        </Text>
        <Image
          ml={2}
          src={Close}
          aria-label='Close'
          alt='Close icon'
        />
      </Box>
    )
  }

  return (
    <Box color='black' bg='white'
      py={2}
      px={4}
      height={10}
      display='flex'
      justifyContent='center'
      borderRadius={4}
      as="button"
      onClick={handleClick}
    >
      <Text fontSize='1rem' lineHeight='24px' textAlign='center'>
        {text}
      </Text>
    </Box>
  );
};

export default Genres;
