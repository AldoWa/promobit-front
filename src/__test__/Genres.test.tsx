import Genres from "../components/Genres"
import { GenresContextData } from "../context/genresContext"
import { screen, render, fireEvent } from "../utils/test"

jest.mock('../services/api.ts', () => {
	return {
    AUTHORIZATION: 'XXXX'
  }
})

describe('Genres', () => {
  const providerProps: GenresContextData = {
    genres: [
      {
        id: 28,
        name: 'Action'
      }
    ],
    handleGenre: () => {},
    genresIDsSelected: [],
    resetGenresSelected: () => {}
  }

  const handleGenre = jest.fn(
    (id: number, type: 'add' | 'remove') => {
      if (type === 'add') {
        providerProps.genresIDsSelected.push(id)
        return;
      }
      providerProps.genresIDsSelected = providerProps.genresIDsSelected.filter(genre => genre !== id)
    }
  ) 

  it('Should render the genre with a text "action"', () => {
    render(<Genres text='Action' id={28} handleGenre={() => {}}/>, {
      providerProps
    })
    const genreComponent = screen.getByTestId('genre')
    expect(genreComponent).toBeInTheDocument()
    expect(genreComponent).toBeVisible()
    
    const genreText = screen.getByTestId('genre-text')
    expect(genreText).toBeInTheDocument()
    expect(genreText).toHaveTextContent('Action')
  })

  it('Should select and deselected when click on the genre', () => {
    render(<Genres text='Action' id={28} handleGenre={handleGenre}/>, {
      providerProps
    })

    expect(handleGenre).not.toHaveBeenCalled()

    const genreComponent = screen.getByTestId('genre')

    const genreComponentSelected = screen.queryByTestId('genre-selected')

    expect(genreComponentSelected).toBeNull()

    fireEvent.click(genreComponent)

    const genreComponentSelectedAfterClick = screen.getByTestId('genre-selected')
    
    expect(genreComponentSelectedAfterClick).toBeInTheDocument()

    fireEvent.click(genreComponent)

    expect(handleGenre).toHaveBeenCalledTimes(2)

    expect(screen.queryByTestId('genre-selected')).not.toBeInTheDocument()
  })

  it('Should match snapshot', () => {
    const { container } = render(<Genres text='Action' id={28} handleGenre={handleGenre}/>, {
      providerProps
    })
    expect(container).toMatchSnapshot()
  })
})