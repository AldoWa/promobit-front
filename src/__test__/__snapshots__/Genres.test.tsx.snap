// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genres Should match snapshot 1`] = `
<div>
  <button
    class="css-1ru2wq5"
    data-testid="genre"
  >
    <p
      class="chakra-text css-1544uh1"
      data-testid="genre-text"
    >
      Action
    </p>
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
